### PR TITLE
Add URL Handler (READY FOR REVIEW)

### PR DESCRIPTION
**Pull request by hqiu-at-wiris**

### CODE REVIEW

File: santoku/salesforce/lightning.py [hunk 1]
```diff
@@ -525,7 +526,7 @@ def do_query_with_SOQL(
         do_request : this method does a request of type GET.
 
         Notes
-        ----
+        -----
         Use this method when you know which objects the data resides in, and you want to
         retrieve data from a single salesforce object or from multiple objects that are related.
```
Comment (dmas-at-wiris) [inline santoku/salesforce/lightning.py]: Salesforce

File: santoku/salesforce/lightning.py [hunk 2]
```diff
@@ -509,6 +509,7 @@ def do_query_with_SOQL(
 
         drop_empty_columns: bool, Optional
             Comlumns that do not contain any value will be removed. Set to false by default.
```
Comment (dmas-at-wiris) [inline santoku/salesforce/lightning.py]: Columns...

File: santoku/utils/url_handler.py [hunk 3]
```diff
@@ -0,0 +1,326 @@
...
+    ) -> str:
+        """
+        Given a URL, return the domain name up to a particular number of subdomains.
+
+        Parameters
+        ----------
+        url : str
+            The URL to get the partial domain from.
+
+        num_subdomains : int
```
Comment (dmas-at-wiris) [inline santoku/utils/url_handler.py]: , Optional

File: santoku/utils/url_handler.py [hunk 4]
```diff
@@ -0,0 +1,326 @@
...
+            Number of subdomains that are extracted.
+
+        return_invalid_urls : bool
```
Comment (dmas-at-wiris) [inline santoku/utils/url_handler.py]: , Optional

File: santoku/utils/url_handler.py [hunk 5]
```diff
@@ -0,0 +1,326 @@
...
+                    return res.domain
+
+                elif res.suffix:
+                    # URL contain only sufix, we return only the suffix
+                    return res.suffix
+
+                # If URL doesn't contain anything identified as domain or suffix, return as it is
+                return url
+
+            return ""
```
Comment (dmas-at-wiris) [inline santoku/utils/url_handler.py]: Shall we raise an exception?

File: santoku/utils/url_handler.py [hunk 6]
```diff
@@ -0,0 +1,326 @@
...
+        """
+        Given a URL return its fully qualified domain name without the trailing dot, defined as the
+        domain name with all its subdomains.
+
+        Parameters
+        ----------
+        url : str
+            The URL to get the fully qualified domain.
+
+        return_invalid_urls : bool
```
Comment (dmas-at-wiris) [inline santoku/utils/url_handler.py]: , Optional

File: santoku/utils/url_handler.py [hunk 7]
```diff
@@ -0,0 +1,326 @@
...
+        -----
+        The verification of IP has been done using the `ipaddress` module of python. More
+        information on the ipaddress module: [1]
+
+        References
+        ----------
+        https://docs.python.org/3/library/ipaddress.html
+
+        """
+        domain = tldextract.extract(url=url).domain
```
Comment (dmas-at-wiris) [inline santoku/utils/url_handler.py:195]: any exception to handle?

**Review by dmas-at-wiris — CHANGES_REQUESTED** [review]
See comments

File: santoku/utils/test/test_url_handler.py [hunk 8]
```diff
@@ -0,0 +1,306 @@
...
+    # Each tuple contains: the raw URL, number of subdomains to extract, the expected partial domain
+    # that the test must return if return_invalid_urls is set to False, expected partial domain if
+    # return_invalid_urls is set to True
+    return [
+        ("https://sub2.sub1.example.com/path?query#fragment", 0, "example.com", "example.com"),
+        ("https://sub2.sub1.example.com", 1, "sub1.example.com", "sub1.example.com"),
+        ("https://sub2.sub1.example.com", 2, "sub2.sub1.example.com", "sub2.sub1.example.com"),
+        ("https://sub2.sub1.example.com", 3, "sub2.sub1.example.com", "sub2.sub1.example.com"),
+        ("https://example.com", 1, "example.com", "example.com"),
+        ("https://example.co.uk", 0, "example.co.uk", "example.co.uk"),
```
Comment (dmas-at-wiris) [inline santoku/utils/test/test_url_handler.py:21]: maybe add a test with port?

**Review by rroig-at-wiris — APPROVED** [review]

File: santoku/utils/url_handler.py [hunk 9]
```diff
@@ -0,0 +1,373 @@
...
+                    return res.domain
+
+                elif res.suffix:
+                    # URL contain only sufix, we return only the suffix
+                    return res.suffix
+
+                # If URL doesn't contain anything identified as domain or suffix, return as it is
+                return url
+
+            raise InvalidURLError(f"The given URL '{url}' does not contain domain or suffix.")
```
Comment (dmas-at-wiris) [inline santoku/utils/url_handler.py]: The {url} URL does not contain domain or suffix

File: santoku/utils/url_handler.py [hunk 10]
```diff
@@ -0,0 +1,373 @@
...
```
Comment (dmas-at-wiris) [inline santoku/utils/url_handler.py]: Same as before

**Review by dmas-at-wiris — APPROVED** [review]
Change those two comments and merge directly. Good job! :)

**Pull request merged**